### PR TITLE
IPC::Attachment is used as an encoded type on Darwin and Windows

### PR DESCRIPTION
**Pull request by kkinnunen-apple**
#### 5540bc1e9de439efe4fbad14560569e807318332
<pre>
IPC::Attachment is used as an encoded type on Darwin and Windows
<a href="https://bugs.webkit.org/show_bug.cgi?id=244882">https://bugs.webkit.org/show_bug.cgi?id=244882</a>
rdar://problem/99638429

Reviewed by Fujii Hironori.

IPC messages were using IPC::Attachment when IPC::Connection was
being transferred via IPC. This was a problem as it is not what
IPC::Attachment is intended to implement. Also the type does not
provide a type safe explanation of the  object being transferred.

IPC::Attachment was encoded and decoded. This was a problem
because IPC::Attachment is the attachment to the encoder or
decoder. It is logically redundant to have encode/decode
capability in the type that has the purpose of representing
something that cannot be encoded or decoded.

Fix by introducing IPC::Connection::Handle type for transferrable
IPC::Connection. Currently that is Win32Handle, UnixFileDescriptor
or MachSendRight.

Remove IPC::Attachment from OS(WINDOWS), as there are no objects
that would be transferred as non-data. Win32Handles are transferred
as normal PID,HANDLE data number pair.

Remove IPC::Attachment::encode(), as each un-encodable object
should add themselves as encoder attachments upon encoding.

Remove IPC::Attachment::decode(), same logic by above.

* LayoutTests/ipc/create-connection-and-send-async.html:
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
(WebKit::GPUConnectionToWebProcess::create):
(WebKit::GPUConnectionToWebProcess::GPUConnectionToWebProcess):
(WebKit::GPUConnectionToWebProcess::createRenderingBackend):
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.h:
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.messages.in:
* Source/WebKit/GPUProcess/GPUProcess.cpp:
(WebKit::GPUProcess::createGPUConnectionToWebProcess):
(WebKit::asConnectionIdentifier): Deleted.
* Source/WebKit/GPUProcess/GPUProcess.h:
* Source/WebKit/GPUProcess/GPUProcess.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::create):
(WebKit::RemoteRenderingBackend::RemoteRenderingBackend):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::createNetworkConnectionToWebProcess):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/Platform/IPC/Attachment.cpp: Removed.
* Source/WebKit/Platform/IPC/Attachment.h:
(IPC::Attachment::customWriter const):
(IPC::Attachment::Attachment::type const): Deleted.
(IPC::Attachment::Attachment::isNull const): Deleted.
(IPC::Attachment::Attachment::size const): Deleted.
(IPC::Attachment::Attachment::fd const): Deleted.
(IPC::Attachment::Attachment::release): Deleted.
(IPC::Attachment::Attachment::customWriter const): Deleted.
* Source/WebKit/Platform/IPC/Connection.h:
(IPC::Connection::Identifier::Identifier):
* Source/WebKit/Platform/IPC/Decoder.cpp:
(IPC::Decoder::takeLastAttachment):
* Source/WebKit/Platform/IPC/ReceiverMatcher.h:
* Source/WebKit/Platform/IPC/StreamServerConnection.cpp:
(IPC::StreamServerConnection::createWithDedicatedConnection):
* Source/WebKit/Platform/IPC/cocoa/ConnectionCocoa.mm:
(IPC::Connection::open):
(IPC::Connection::sendOutgoingMessage):
(IPC::Connection::receiveSourceEventHandler):
(IPC::Connection::createConnectionIdentifierPair):
* Source/WebKit/Platform/IPC/darwin/ArgumentCodersDarwin.h:
* Source/WebKit/Platform/IPC/darwin/ArgumentCodersDarwin.mm:
(IPC::ArgumentCoder&lt;MachSendRight&gt;::encode):
(IPC::ArgumentCoder&lt;MachSendRight&gt;::decode):
* Source/WebKit/Platform/IPC/unix/AttachmentUnix.cpp:
(IPC::Attachment::Attachment):
(IPC::Attachment::encode const):
(IPC::Attachment::decode):
* Source/WebKit/Platform/IPC/unix/ConnectionUnix.cpp:
(IPC::Connection::createConnectionIdentifierPair):
* Source/WebKit/Platform/IPC/win/AttachmentWin.cpp: Removed.
* Source/WebKit/Platform/IPC/win/ConnectionWin.cpp:
(IPC::Connection::createConnectionIdentifierPair):
* Source/WebKit/Platform/Sources.txt:
* Source/WebKit/PlatformWin.cmake:
* Source/WebKit/Scripts/webkit/messages.py:
(argument_coder_headers_for_type):
* Source/WebKit/Shared/IPCConnectionTester.cpp:
(WebKit::IPCConnectionTester::create):
(WebKit::IPCConnectionTester::IPCConnectionTester):
(WebKit::asIdentifier): Deleted.
* Source/WebKit/Shared/IPCConnectionTester.h:
* Source/WebKit/Shared/IPCTester.cpp:
(WebKit::IPCTester::createConnectionTester):
(WebKit::IPCTester::createConnectionTesterAndSendAsyncMessages):
* Source/WebKit/Shared/IPCTester.h:
* Source/WebKit/Shared/IPCTester.messages.in:
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
(WebKit::GPUProcessProxy::createGPUProcessConnection):
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.h:
* Source/WebKit/UIProcess/Inspector/WebInspectorUIProxy.cpp:
(WebKit::WebInspectorUIProxy::setFrontendConnection):
* Source/WebKit/UIProcess/Inspector/WebInspectorUIProxy.h:
* Source/WebKit/UIProcess/Inspector/WebInspectorUIProxy.messages.in:
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::createGPUProcessConnection):
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::createGPUProcessConnection):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/UIProcess/WebProcessProxy.messages.in:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::getNetworkProcessConnection):
* Source/WebKit/UIProcess/linux/MemoryPressureMonitor.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Inspector/WebInspector.cpp:
(WebKit::WebInspector::setFrontendConnection):
* Source/WebKit/WebProcess/Inspector/WebInspector.h:
* Source/WebKit/WebProcess/Inspector/WebInspector.messages.in:
* Source/WebKit/WebProcess/Network/NetworkProcessConnection.h:
(WebKit::NetworkProcessConnection::create):
* Source/WebKit/WebProcess/Network/NetworkProcessConnectionInfo.h:
(WebKit::NetworkProcessConnectionInfo::identifier const): Deleted.
(WebKit::NetworkProcessConnectionInfo::releaseIdentifier): Deleted.
* Source/WebKit/WebProcess/WebPage/IPCTestingAPI.cpp:
(WebKit::IPCTestingAPI::JSIPCConnectionHandle::create):
(WebKit::IPCTestingAPI::JSIPCConnectionHandle::encode const):
(WebKit::IPCTestingAPI::JSIPCConnectionHandle::JSIPCConnectionHandle):
(WebKit::IPCTestingAPI::JSIPCConnectionHandle::createJSWrapper):
(WebKit::IPCTestingAPI::JSIPCConnectionHandle::wrapperClass):
(WebKit::IPCTestingAPI::JSIPCConnectionHandle::unwrap):
(WebKit::IPCTestingAPI::JSIPCConnectionHandle::toWrapped):
(WebKit::IPCTestingAPI::JSIPCConnectionHandle::initialize):
(WebKit::IPCTestingAPI::JSIPCConnectionHandle::finalize):
(WebKit::IPCTestingAPI::JSIPCConnectionHandle::staticFunctions):
(WebKit::IPCTestingAPI::encodeConnectionHandle):
(WebKit::IPCTestingAPI::encodeArgument):
(WebKit::IPCTestingAPI::JSIPC::createConnectionPair):
(WebKit::IPCTestingAPI::JSIPCAttachment::create): Deleted.
(WebKit::IPCTestingAPI::JSIPCAttachment::encode const): Deleted.
(WebKit::IPCTestingAPI::JSIPCAttachment::JSIPCAttachment): Deleted.
(WebKit::IPCTestingAPI::JSIPCAttachment::createJSWrapper): Deleted.
(WebKit::IPCTestingAPI::JSIPCAttachment::wrapperClass): Deleted.
(WebKit::IPCTestingAPI::JSIPCAttachment::unwrap): Deleted.
(WebKit::IPCTestingAPI::JSIPCAttachment::toWrapped): Deleted.
(WebKit::IPCTestingAPI::JSIPCAttachment::initialize): Deleted.
(WebKit::IPCTestingAPI::JSIPCAttachment::finalize): Deleted.
(WebKit::IPCTestingAPI::JSIPCAttachment::staticFunctions): Deleted.
(WebKit::IPCTestingAPI::encodeAttachment): Deleted.
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::getNetworkProcessConnection):
(WebKit::WebProcess::ensureNetworkProcessConnection):
* Source/WebKit/WebProcess/cocoa/PlaybackSessionManager.h:

Canonical link: <a href="https://commits.webkit.org/254379@main">https://commits.webkit.org/254379@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/82d562fe7e23cb2f50bc3073d84b4fb8cb2c5edb

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/88687 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/33249 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/19570 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/97876 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/154424 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/31742 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/27357 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/80909 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/92512 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/94318 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/25188 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/75675 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/25125 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/80070 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/80136 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/68090 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/29501 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/14142 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/29267 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/15148 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3071 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/32713 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/38083 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/31396 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/34259 "Passed tests") | |

### CODE REVIEW

**Comment by fujii** [comment]
> IPC messages were using IPC::Attachment when IPC::Connection was
> being transferred via IPC. This was a problem as it is not what
> IPC::Attachment is intended to implement. Also the type does not
> provide a type safe explanation of the  object being transferred.
> 
> IPC::Attachment was encoded and decoded. This was a problem
> because IPC::Attachment is the attachment to the encoder or
> decoder. It is logically redundant to have encode/decode
> capability in the type that has the purpose of representing
> something that cannot be encoded or decoded.

`IPC::Attachment` is a adopter converting a unencodable connection identifier to an encodable object.
It think it's the original author's intension.
It looks like your change just renames `IPC::Attachment` to `IPC::Connection::Handle`.
BTW, why don't you remove `IPC::Attachment` completely?

**Comment by kkinnunen-apple** [comment]
Thanks for looking!

> It looks like your change just renames IPC::Attachment to IPC::Connection::Handle.

The change here changes IPC:Attachment used as connection handle to the platform file handle type.
Before: IPC::Connection handle was IPC::Attachment
After: IPC::Connection handle is the platform file handle type.

In other words:
* IPC::Connection handle is not a IPC::Attachment.
* IPC::Connection handle is a file handle or a mach port.
* Mach port is transferred as IPC::Attachment.
* File handle is transferred as IPC::Attachment on unix.

> BTW, why don't you remove IPC::Attachment completely?

Because IPC::Attachment in original authors' view represented "attachment to the IPC message data, representing objects that cannot be transferred as message data." There are still such things: Mach ports and Unix file handles. This patch does not change this design.

This patch just removes the incorrect "encode" and "decode" operation.

File: Source/WebKit/Platform/IPC/unix/AttachmentUnix.cpp [hunk 1]
```diff
@@ -76,4 +81,14 @@ Attachment& Attachment::operator=(Attachment&& attachment)
 
 Attachment::~Attachment() = default;
 
+void Attachment::encode(Encoder& encoder) const
```
Comment (fujii) [inline Source/WebKit/Platform/IPC/unix/AttachmentUnix.cpp]: I think Attachment::encode and Attachment::decode should be removed in your design.

Comment (kkinnunen-apple) [reply Source/WebKit/Platform/IPC/unix/AttachmentUnix.cpp]: Yes, but the other patch does this for unix. See pr #4122.
This patch removes them only for Darwin and Windows.

File: Source/WebKit/Platform/IPC/Connection.h [hunk 2]
```diff
@@ -132,23 +141,43 @@ class Connection : public ThreadSafeRefCounted<Connection, WTF::DestructionThrea
     };
 #endif
 
+#if USE(UNIX_DOMAIN_SOCKETS)
+    using Handle = UnixFileDescriptor;
+#elif OS(WINDOWS)
+    using Handle = Win32Handle;
+#elif OS(DARWIN)
+    using Handle = MachSendRight;
+#endif
```
Comment (fujii) [inline Source/WebKit/Platform/IPC/Connection.h]: You removed SharedMemory::IPCHandle. https://bugs.webkit.org/show_bug.cgi?id=240056
Connection::Handle is like another IPCHandle. Can you remove it?

Comment (kkinnunen-apple) [reply Source/WebKit/Platform/IPC/Connection.h]: No.
SharedMemory needs a handle type (SharedMemory::Handle) to serialise SharedMemory reference to a handle.
Connection needs a handle type (Connection::Handle) to serialise Connection reference to a handle.

Comment (fujii) [reply Source/WebKit/Platform/IPC/Connection.h]: Does Connection need both Handle and Identifier?

Comment (kkinnunen-apple) [reply Source/WebKit/Platform/IPC/Connection.h]: Currently yes. #4059 is working towards removing Connection::Identifier

**Review by fujii — APPROVED** [review]
I don't agree with the descriptions of your commit message. But, I don't object to your new design.

**Comment by kkinnunen-apple** [comment]
> I don't agree with the descriptions of your commit message. But, I don't object to your new design.

Thanks!
Which part do you not agree?

Which part do you think is new?

**Comment by fujii** [comment]
IMO, `IPC::Attachment` is an adapter converting a unencodable connection identifier to an encodable object. It isn't a problem encoding/decoding `IPC::Attachment`.

**Comment by kkinnunen-apple** [comment]
> IMO, `IPC::Attachment` is an adapter converting a unencodable connection identifier to an encodable object. It isn't a problem encoding/decoding `IPC::Attachment`.

You cannot convert an unencodable object (file handle on unix, mach port on Mac) to an encodable object. If there was a way, there would not be any undecodable objects. There's no way to do that. If you have file handle 0x7774747, there is no way to transform that piece of 64-bit data to another piece of 64-bit data that would work in another process.

The way to do that is to send that with the operating system kernel send primitives.
The way this is abstracted in WebKit IPC is to add a reference to the object inside a IPC::Attachment and add that as an attachment to the encoder. Then when the message is sent, the kernel send primitive to said object is used.

**Comment by fujii** [comment]
Attachment has encode/decode methods since the beginning.
https://github.com/WebKit/WebKit/blob/cde8ff54ceb2eb4491dc879eca1aa3824fdff228/WebKit2/Platform/CoreIPC/Attachment.h#L67-L68
It's better to ask the auther the original intention.

**Comment by kkinnunen-apple** [comment]
Would that be more of an artefact that there were no rvalue types at that time and the mach ports didn't have suitable abstraction. I'm not sure the original author is available anymore (not sure Sam wrote the code even though he committed it). I can ping Sam but I'd not like to block on this.

I think the logical arguments here should still hold, or what do you think?
E.g.:

- Attachment is an attachment to the Encoder/Decoder
- No object is ever expected to interact with Attachment. E.g. connection does not hold or return IPC::Attachment, rather it returns its handle (of its handle type).

Example of Connection::Handle: it is a class useful in and of itself, regardless whether the handle is sent to another process. It can be used in-process also. It does not make sense to use IPC::Attachment, as that conveys nothing about what the object actually is.

**Comment by fujii** [comment]
Your patch is good. I dont mean to block your work. Feel free to land.

**Pull request closed without merging**